lib/vecmat_external.h

```diff
@@ -40,48 +40,167 @@
+}
 };
-
-#define IDENTITY_MATRIX                                                                                                \
-  {                                                                                                                    \
-    {1.0, 0, 0}, {0, 1.0, 0}, { 0, 0, 1.0 }                                                                            \
-  }
 
 struct vector {
-  float x, y, z;
+using T = scalar;
+constexpr static const size_t N = 3;
+constexpr static const size_t X = 0;
+constexpr static const size_t Y = 1;
+constexpr static const size_t Z = 2;
+union {
+  T xyz[N];
+  struct { T x, y, z; };
+  struct { T r, g, b; };
+};
+constexpr static inline const vector id(ssize_t i = -1)
+{
+  return (vector){
+  (scalar)(i % N == X ? 1 : 0),
+  (scalar)(i % N == Y ? 1 : 0),
+  (scalar)(i % N == Z ? 1 : 0)
+  };
+}
+constexpr static inline const vector ne()
+{
+	return (vector){ (scalar)0, (scalar)0, (scalar)0 };
+}
+};
+
+using vector_array = vector;
+
+struct alignas(sizeof(scalar) * 4) aligned_vector {
+using T = scalar;
+constexpr static const size_t N = 3;
+constexpr static const size_t X = 0;
+constexpr static const size_t Y = 1;
+constexpr static const size_t Z = 2;
+union {
+  T xyz[N];
+  struct { T x, y, z; };
+  struct { T r, g, b; };
+};
+constexpr static inline const aligned_vector id(ssize_t i = -1)
+{
+  return (aligned_vector){
+  (scalar)(i % N == X ? 1 : 0),
+  (scalar)(i % N == Y ? 1 : 0),
+  (scalar)(i % N == Z ? 1 : 0)
+  };
+}
+constexpr static inline const aligned_vector ne()
+{
+	return (aligned_vector){ (scalar)0, (scalar)0, (scalar)0 };
+}
 };
 
-struct vector4 {
-  float x, y, z, kat_pad;
+using aligned_vector_array = aligned_vector;
+
+struct alignas(sizeof(scalar) * 4) vector4 {
+using T = scalar;
+constexpr static const size_t N = 4;
+constexpr static const size_t X = 0;
+constexpr static const size_t Y = 1;
+constexpr static const size_t Z = 2;
+constexpr static const size_t W = 3;
+union {
+  T xyzw[N];
+  struct { T x, y, z; union { T w; T kat_pad; }; };
+  struct { T r, g, b, a; };
 };
 
-struct vector_array {
-  float xyz[3];
+constexpr static inline const vector4 id(ssize_t i = -1)
+{
+  return (vector4){
+  (scalar)(i % N == X ? 1 : 0),
+  (scalar)(i % N == Y ? 1 : 0),
+  (scalar)(i % N == Z ? 1 : 0),
+  (scalar)(i % N == W ? 1 : 0)
+  };
+}
+constexpr static inline const vector4 ne()
+{
+	return (vector4){ (scalar)0, (scalar)0, (scalar)0, (scalar)0 };
+}
 };
 
+using vector4_array = vector4;
+using aligned_vector4 = vector4;
+using aligned_vector4_array = aligned_vector4;
+
 struct matrix {
-  vector rvec, uvec, fvec;
+  vector rvec;
+  vector uvec;
+  vector fvec;
+constexpr static inline const matrix id()
+{
+  return (matrix){
+  vector::id(0),
+  vector::id(1),
+  vector::id(2)
+  };
+}
+constexpr static inline const matrix ne()
+{
+	return (matrix){ vector::ne(), vector::ne(), vector::ne() };
+}
 };
 
+constexpr const matrix IDENTITY_MATRIX = matrix::id();
```

Comment:
this should be the same as `constexpr matrix IDENTITY_MATRIX = matrix::id();`

Reply:
yes you are right constexpr implies const.